File: erlang/pe002.erl
-module(pe002).
-export([start/0]).

maxv() -> 4000000.

start() ->
  pe002(0, 1, 0, maxv()).
  
pe002(_Prev, Current, Sum, Maxv) when (Current >= Maxv) ->
    io:format("~w~n", [Sum]);
    
pe002(Prev, Current, Sum, Maxv) when (Current rem 2 == 0) ->
    pe002(Current, Prev+Current, Sum+Current, Maxv);
    
pe002(Prev, Current, Sum, Maxv) ->
    pe002(Current, Prev+Current, Sum, Maxv).

File: ocaml/pe002.ml
let even x = x mod 2 = 0 ;;

let rec pe002 n previous current sum =
  if current > n then
    Printf.printf "%d\n" sum
  else
    if even current then 
      pe002 n current (previous + current) (sum + current)
    else 
      pe002 n current (previous + current) sum;;

let main () =
  pe002 4_000_000 1 1 0

let () = main ()
